Technical performance improvements for 74x

davidlange6 opened the pull request:
1) Don't make a new cut parser object for every jet and lepton (do it once instead)
2) deltaR2(candidate,candidate) is much faster than deltaR2(candidate.p4(),candidate.p4())
3) remove memory leaks from HypDilepMaker
4) reserve memory and move if statements out of loop in de/dx producer

matteosan1 committed:
add fillDescriptions to EcalUncalibRechitProducer

matteosan1 committed:
cleaned producer

matteosan1 committed:
use ifValue

matteosan1 committed:
added default config

matteosan1 committed:
final version

matteosan1 committed:
fix in pset name

matteosan1 committed:
fix of default cfi

matteosan1 committed:
finally fixed multifit config

Martin-Grunewald committed:
Update for Ecal/fillDescription/MultiFit

Commit:
plot bx for all triggers

jdolen committed:
adding softdrop angular normalization parameter

ferencek committed:
moved soft lepton TagInfos and taggers under the candidate-based fwk

ferencek committed:
switching to the old parameter name for the PV collection

ferencek committed:
added MiniAOD support for soft lepton taggers

perrotta committed:
Muon HLT paths for the HIon menu, CMSHLT-198

ferencek committed:
fixed inverted conversion veto condition

Martin-Grunewald committed:
CMSHLT-255 HLTL1UnpackerSequence

jdolen committed:
fixing R0 value for ak8PFJetsCHSSoftDrop

Commit:
clean up some useless headers

ktf committed:
Drop a bunch of unneeded dependencies on rootrflx

Since DataFormats/DetId depends on it, basically everything else does.

davidlange6 committed:
remove more printouts

davidlange6 committed:
catch failure mode at start of function instead of end

davidlange6 committed:
memory leaks and perf improvements

davidlange6 committed:
Merged pre8_perf from repository davidlange6

andrewj314 committed:
Bugfix in TauReco/src/classes_def_2.xml adding ioread rules to clear transient vectors

https://hypernews.cern.ch/HyperNews/CMS/get/physTools/3315/1/1.html

andrewj314 committed:
Transient Vector Bugfix

davidlange6 committed:
remove printout i added by mistake

davidlange6 committed:
big speedups in singletop dqm and ipproducer

cmsbuild commented:
A new Pull Request was created by @davidlange6 (David Lange) for CMSSW_7_4_X.

Technical performance improvements for 74x

It involves the following packages:

DQM/Physics
DQM/PhysicsHWW
RecoBTag/ImpactParameter
RecoTauTag/RecoTau
RecoTracker/DeDx


@nclopezo, @cvuosalo, @slava77, @cmsbuild, @deguio, @danduggan can you please review it and eventually sign? Thanks.
@ghellwig, @rappoccio, @imarches, @makortel, @acaudron, @jpavel, @GiacomoSguazzoni, @rovere, @VinInn, @appeltel, @ferencek, @gpetruc, @cerati, @pvmulder, @dgulhan, @venturia this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

deguio committed:
working in production by default

patch to be applied in case of playback deployement

patch to be applied in case of user area deployement

changing the default to users gui

adding patch for prod installation

andrewj314 committed:
bugfix fourth attempt

VinInn commented on DQM/PhysicsHWW/src/HypDilepMaker.cc:
```diff
@@ -122,35 +116,35 @@ void HypDilepMaker::SetVars(HWW& hww, const edm::Event& iEvent, const edm::Event
 	
         // we don't want jets that overlap with electrons
         bool overlapsWithLepton = false;
-        if(!testJetForLeptons(jets_p4->at(i), mus_p4->at(loose_index))) 
```
given the index of the loop is hyper safe to use [i] instead of .at(i)
``std::vector<T>::at(..)'' should be banned.
I suggest to trap its use in some code analyzer

VinInn commented on RecoTauTag/RecoTau/plugins/PFRecoTauChargedHadronFromTrackPlugin.cc:
```diff
@@ -143,13 +143,14 @@ PFRecoTauChargedHadronFromTrackPlugin::return_type PFRecoTauChargedHadronFromTra
-    double dR = deltaR(track->eta(), track->phi(), jet.eta(), jet.phi());
+    double dR = deltaR(track->eta(), track->phi(), jEta,jPhi);
     double dRmatch = dRcone_;
     if ( dRconeLimitedToJetArea_ ) {
```
while factorizing also this part can be moved outside the loop...
and (deltaR moved to deltaR2)
(we should ban deltaR)

davidlange6 replied:
interesting…

deltaR is everywhere…

Is there a (twiki) summary of deltaR* tricks? I’m pretty lost myself just looking at the code

> On Mar 4, 2015, at 1:34 PM, Vincenzo Innocente <notifications@github.com> wrote:
> 
> In RecoTauTag/RecoTau/plugins/PFRecoTauChargedHadronFromTrackPlugin.cc:
> 
> >    size_t numTracks = tracks->size();
> >    for ( size_t iTrack = 0; iTrack < numTracks; ++iTrack ) {
> >      reco::TrackRef track(tracks, iTrack);
> >  
> >      // consider tracks in vicinity of tau-jet candidate only
> > -    double dR = deltaR(track->eta(), track->phi(), jet.eta(), jet.phi());
> > +    double dR = deltaR(track->eta(), track->phi(), jEta,jPhi);
> >      double dRmatch = dRcone_;
> >      if ( dRconeLimitedToJetArea_ ) {
> 
> while factorizing also this part can be moved outside the loop...
> and (deltaR moved to deltaR2)
> (we should ban deltaR)
> 
> —
> Reply to this email directly or view it on GitHub.
>

VinInn replied:
no, maybe we should compile one and provide some clever algorithm to perform NN searches in eta-phi.
The problem is, as with any combinatorics, that the optimal algorithm depends on the complexity of the problem
(see fastjet).
In principle one could use fastjet each time deltaR is used...  (or PF algos)

lgray replied:
Fastjet without computational geometry package included in the build is still n^2, so won't help. 
Probably kd-tree + neighbours cache would fix this problem.

VinInn commented on DQM/PhysicsHWW/src/HypDilepMaker.cc:
```diff
@@ -48,33 +48,27 @@ void HypDilepMaker::SetVars(HWW& hww, const edm::Event& iEvent, const edm::Event
   // muon charge
-  vector<int> *mus_charge = new vector<int>;
-  *mus_charge = hww.mus_charge(); 
+  vector<int> mus_charge = hww.mus_charge(); 
```
wonder if all these assignments should not be just
const auto & bha = hww.bha();
(was this a CINT macro turned into CMSSW producer or what?)

davidlange6 replied:
indeed - i was too lazy to check if that would work or not yesterday..

> On Mar 4, 2015, at 1:41 PM, Vincenzo Innocente <notifications@github.com> wrote:
> 
> In DQM/PhysicsHWW/src/HypDilepMaker.cc:
> 
> > @@ -48,33 +48,27 @@ void HypDilepMaker::SetVars(HWW& hww, const edm::Event& iEvent, const edm::Event
> >    double hypJetMinPtCut = 30.0;
> >  
> >    // muon charge
> > -  vector<int> *mus_charge = new vector<int>;
> > -  *mus_charge = hww.mus_charge(); 
> > +  vector<int> mus_charge = hww.mus_charge(); 
> 
> wonder if all these assignments should not be just
> const auto & bha = hww.bha();
> (was this a CINT macro turned into CMSSW producer or what?)
> 
> —
> Reply to this email directly or view it on GitHub.
>

davidlange6 replied:
ok, the HWW class is already a big common block. So yes, it works fine.

> On Mar 4, 2015, at 2:03 PM, David Lange <David.Lange@cern.ch> wrote:
> 
> indeed - i was too lazy to check if that would work or not yesterday..
> 
>> On Mar 4, 2015, at 1:41 PM, Vincenzo Innocente <notifications@github.com> wrote:
>> 
>> In DQM/PhysicsHWW/src/HypDilepMaker.cc:
>> 
>>> @@ -48,33 +48,27 @@ void HypDilepMaker::SetVars(HWW& hww, const edm::Event& iEvent, const edm::Event
>>>   double hypJetMinPtCut = 30.0;
>>> 
>>>   // muon charge
>>> -  vector<int> *mus_charge = new vector<int>;
>>> -  *mus_charge = hww.mus_charge(); 
>>> +  vector<int> mus_charge = hww.mus_charge(); 
>> 
>> wonder if all these assignments should not be just
>> const auto & bha = hww.bha();
>> (was this a CINT macro turned into CMSSW producer or what?)
>> 
>> —
>> Reply to this email directly or view it on GitHub.
>> 
>

davidlange6 committed:
change at to []

davidlange6 committed:
and avoid a copy of vectors

slava77 commented on RecoTauTag/RecoTau/plugins/PFRecoTauChargedHadronFromTrackPlugin.cc:
```diff
@@ -143,13 +143,14 @@ PFRecoTauChargedHadronFromTrackPlugin::return_type PFRecoTauChargedHadronFromTra
 
     // consider tracks in vicinity of tau-jet candidate only
-    double dR = deltaR(track->eta(), track->phi(), jet.eta(), jet.phi());
+    double dR = deltaR(track->eta(), track->phi(), jEta,jPhi);
```
isn't taking out an obvious const x.func() call out of the loop just a syntactic improvement at this point?
I played a bit in the past with some simple compile units with simple functions and then looking at assembly instructions. Even O1 already takes out repeated function calls outside of the loop.

VinInn replied:
in principle yes, unless the heap and call to external functions are involved and gcc decides to play safe...

rappoccio committed:
Stream migration for JME HLT modules

deguio committed:
fix

Dr15Jones committed:
Make char arrays fully const in TaggingVariable

The globally available arrays of names are now fully const. This
was spotted by the static analyzer.

cmsbuild commented:
+1
Tested at: 98de7ba9a40e1edf9af0dc36b37b15baf7c87c9d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8055/3108/summary.html

cmsbuild commented:
Pull request #8055 was updated. @danduggan, @cvuosalo, @deguio, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8055/3108/summary.html

cmsbuild commented:
+1
Tested at: 98de7ba9a40e1edf9af0dc36b37b15baf7c87c9d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8055/3118/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8055/3118/summary.html

Martin-Grunewald committed:
Remove case FULL from addOn tests

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/8055	98de7ba

 tested in CMSSW_7_5_X_2015-03-04-0200 /sign524/ with a longer version of the matrix run by jenkins.

There are no changes in monitored quantities as expected from changes in this code.

Looking at timing in 4.53 workflow (this takes about 30% of the reco time that we usually check for RECO performance in 25202) the following observations are made.

Among the modules affected by the code changes only the following shows a significant change in wall-clock timing in the reco sequence
```
        pfImpactParameterTagInfos        15.5679 ms/ev -> 8.66813 ms/ev
```

The others (tau and dedx code) have a change in time within the reproducibility jitter.
```
        ak4PFJetsRecoTauChargedHadrons   11.2055 ms/ev -> 11.5193 ms/ev
        dedxTruncated40          5.71675 ms/ev -> 5.67303 ms/ev
        dedxHarmonic2    3.99958 ms/ev -> 3.93844 ms/ev
        dedxHitInfo      0.166835 ms/ev -> 0.167116 ms/ev
```

arizzi committed:
persistently store a valuemap with the scores used for sorting

ktf committed:
Add extra dictionaries as requested by ROOT6.

dmitrijus commented on DQM/Physics/src/SingleTopTChannelLeptonDQM.h:
```diff
@@ -201,6 +201,18 @@ class MonitorEnsemble {
+  
+  StringCutObjectSelector<reco::PFCandidate, true> *elecSelect;
+  StringCutObjectSelector<reco::PFCandidate, true> *elecIso;
+
```
Can we make those to be unique_ptr?
For the sake of completeness and cleanliness.

arizzi committed:
fix association filling

wmtan committed:
Root6 version numbers should be known to Root5. (Core)

davidlange6 committed:
Merge pull request #8090 from ktf/drop-unneeded-dependencies

Drop a bunch of unneeded dependencies on rootrflx

davidlange6 committed:
Merge pull request #8010 from davidlange6/printouts_r2

Remove some more printouts from RECO step

davidlange6 committed:
Merge pull request #8100 from ktf/more-root6-dictionaries

Add extra dictionaries as requested by ROOT6.

cmsbuild committed:
Merge pull request #8078 from Dr15Jones/makeArrayOfCharsConstInTaggingVariable

Make char arrays fully const in TaggingVariable

cmsbuild committed:
Merge pull request #8074 from deguio/patchesForOnline_74

Patches for online

cmsbuild committed:
Merge pull request #8061 from cms-tau-pog/transient-vector-bugfix-secondattempt

Transient vector bugfix

cmsbuild committed:
Merge pull request #7915 from cms-tsg-storm/branch50nsHltMenu74X

HLT update including Ecal for 74X

cmsbuild committed:
Merge pull request #7978 from jdolen/add-softdrop-normalization-parameter

Add softdrop normalization parameter

cmsbuild committed:
Merge pull request #8102 from arizzi/from-CMSSW_7_4_0_pre8

PrimaryVertex sorter fix and store of the score vector - 74X version

cmsbuild committed:
Merge pull request #7956 from jingyucms/jingyu-74x-plot-bx

plot bx for all triggers

davidlange6 committed:
Merge pull request #7959 from cms-btv-pog/PATBTaggingUpdates_from-CMSSW_7_4_0_pre7

Soft lepton tagger updates (74X)

cmsbuild committed:
Merge pull request #8076 from rappoccio/JMEHLTStreamMigration

Stream migration for JME HLT modules (backport to 74x)

cmsbuild committed:
Merge pull request #8110 from matteosan1/topic_fix_rechitinregions_74X

fix rechitinregions

cmsbuild committed:
Merge pull request #8104 from wmtan/ChecksumsInCore

Root6 version numbers should be known to Root5. (Core)

cmsbuild committed:
Merge pull request #8085 from Martin-Grunewald/KillFull74X

Remove case FULL from addOn HLT tests (74X)

deguio commented:
+1
this should go to 75x as well I guess

davidlange6 committed:
Merged refs/pull/8055/head from repository cms-sw

davidlange6 committed:
switch to uniqueptrs

slava77 commented:
we signed 
98de7ba with "+134 −102 in 7 files" changed
github now shows "+12,528 −20,931 in 332 files"

what was actually merged?
should we expect the diff on github to still make sense after the merge happened?

davidlange6 commented:
i added a commit starting from a newer IB.. so the PR picked up the diffs.

> On Mar 11, 2015, at 10:32 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> we signed 
> 98de7ba with "+134 −102 in 7 files" changed
> github now shows "+12,528 −20,931 in 332 files"
> 
> what was actually merged?
> should we expect the diff on github to still make sense after the merge happened?
> 
> —
> Reply to this email directly or view it on GitHub.
>
